### ISSUE TITLE
flux queue start|stop does not work with named queues

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#4629]
Problem: after #4627, `flux queue disable --queue=NAME` works to prevent job submission on a queue basis.  It would be helpful to also allow `flux queue stop --queue=NAME` to prevent jobs in a particular queue from being scheduled.

One use case for this would be allowing jobs to be submitted to an `all` partition in advance of a DAT.

Edit: this administrative state should be preserved across instance restart like in #4628

**Comment by chu11** [comment]
hacked away on a prototype for the flue queue start/stop with named queues as it took a little time to mentally grasp what was going on.  some refactoring steps that maybe should be done:

- should consider splitting up `alloc-admin` RPC into an `alloc-admin` and `alloc-query` or something.  Logically I think makes things more complicated than probably want.  Would lead to an extra rpc call in `flux-queue` commands I think.

- how to format of the response from `alloc-admin` (or future `alloc-query`)?  I temporarily decided to follow what I did in `job-list` with job stats.  I added a `queues` array into the response that will list enable/disable and reason for each queue, if named queues exist.  I think this generally works, although the "reason" field makes no sense if named queues exist (I just put in reasons like "see queue specifics").

- i copied the `jobq` data structure from `job-manager/queue` into `job-manager/alloc`.  Could be made into a util library or something maybe?  Thinking about it.

- the `job-manager` "alloc queue" could probably use a rename, b/c with "job queue" support, code could get confusing.  Although no alternate name came to mind that would be easier.  maybe "schedqueue" or "schedq"?  Still thinking about it.

**Comment by chu11** [comment]
> I could be missing some details but I was hoping we could just add a started flag to jobq in queue.c. There's already a data structure in there that handles one anon queue vs N named queues and saving/restoring state to the KVS. I would have thought the "admin" rpc for start/stop could be patterned after the enable/disable one that's already there.

Yup, in my prototype I largely patterned it off the enable/disable from the `queue-admin`.   Are you suggesting we move `alloc-admin` into `queue.c`?  I just didn't think to do it that way.

```
    stop by scanning the alloc queue for jobs with that queue name and remove them from the queue
    start by scanning the active job hash for jobs with that queue name that are not in the alloc queue and add them
```

that's pretty much what i've done, except in the stop case we also should cancel pending jobs that have been sent to the scheduler.  We don't want those to be allocated resources, either, correct?

**Comment by garlick** [comment]
Yeah, that was what I was thinking - implement at least the RPC handlers and the start/stop state in queue.c.   Then maybe  functions for moving jobs on/off of the one queue could be built in alloc.c and called from queue.c.

Right on getting jobs back from the scheduler that have not yet been allocated resources.  That's what I meant by syncing with the scheduler.   In liimited mode, only the top N priority jobs are exposed to the scheduler so potentially some jobs may need to be gotten back when a queue is started too.

**Comment by chu11** [comment]
@garlick what do you think about modifying the current `queue-admin` RPC to had a `start` boolean?  i.e. have `queue-admin` handle both enable/disable and start/stop?

**Comment by chu11** [comment]
> Then maybe the alloc-admin RPC can be retired.

Oh, and I guess what you mean is `alloc-admin` would regress into just a `alloc-query` or so?  Or are you thinking the single RPC should still return all the alloc info?

**Comment by garlick** [comment]
Yup that's pretty much what I was thinking.  Of course details sometimes get in the way :-)

**Comment by chu11** [comment]
continuing to prototype away building off of the code in `queue.c`, fixing gotchas along the way.

thing I dislike, circular dependency - alloc & queue circularly depend on each other
- queue needs to call alloc functions, enqueue, cancel, recalc prio, etc.
- alloc needs to determine if queue is started/stopped (e.g. when re-queuing)
haven't thought if this is refactorable yet or other ways around it yet, the re-queuing could maybe be handled via  flag or something in the job struct?

if all queues disabled (including if its just the anonymous one), the `alloc->queue` will be empty, which is a change in behavior.  this ok?

current devel @ https://github.com/chu11/flux-core/tree/issue4629_queue_start_stop  - but its messy

**Comment by chu11** [comment]
thinking about it a bit more, prior to this work, `alloc` sort of lived and did all of its stuff on its own and `queue` did the same, so `queue` calling into `alloc` and vice versa sorta makes this all confusing.  I think mentally when I had separate "jobq" data structures inside `alloc.c`, it was mentally easier.
- Perhaps registering "callbacks" in the `queue.c`  would be wiser.  So `alloc.c` would simply register callbacks and all logic is maintained inside `alloc.c` a bit more.

alternate idea, I think originally we decided to do this the "brute force" way and maintain just the single queue like before.  Perhaps moving to maintaining separate named queues inside alloc might be easier afterall?

**Comment by garlick** [comment]
I haven't looked in detail yet (and may not get to for a bit) but I would say that a goal  for now should be to touch the alloc.c code as little as possible since that code is critical and this feature is not a really high priority AFAIK.  If we can build the new functionality into queue.c and just add a few helpers in alloc.c, that would be ideal.  I am fine with moving the admin RPC out of alloc.c though.

It doesn't bother me too much if alloc.c and queue.c are a little bit co-dependent, although if the complexity is not low then we should take another look at whether this hack is worth it, and maybe consider  "queues for real" as you suggest, although I think that would be a much bigger job.

**Comment by chu11** [comment]
although it could be for a separate issue, any reason we didn't have queue support for drain/idle in this issue?  Seems like it should follow.

**Comment by garlick** [comment]
Probably a separate issue since this issue is tracking a piece of missing functionality that we need and I'm not sure if there is a use case to drive per-queue drain/idle.
